allennlp/nlp_api.py

```diff
@@ -0,0 +1,289 @@
+from typing import Callable, Dict, List, Union
+
+import torch
+
+from .common.checks import ConfigurationError
+from .common import Params
+
+
+class NlpApi:
+    """
+    The ``NlpApi`` provides a high-level API for building NLP models.  You don't have to use it
+    for building a model with AllenNLP, but we think you should, because it's useful.
+
+    ``NlpApi`` abstracts away particular decisions that are frequently necessary in NLP, like how
```

Comment:
+1 to Christian's comment about making data flow explicit.  Assuming we provide good templates, I think it's better to remove this bit of magic as it makes it explicit what is happening to the user and easier to understand without knowing the lower level details about how the class works.  The current method of parsing the fallback behavior string and chaining the defaults is difficult to immediately understand and this may provide a barrier to others adopting the code.